### ISSUE TITLE
Can't seem to get Cordova Media to work in Capacitor & Ionic React

### ISSUE THREAD

**Issue by d9media** [apache/cordova-plugin-media#265]
I am currently trying to get the plug-in to work with capacitor and Ionic react.

First, I've installed the plug-in as suggested running the following commands described on this help page: https://capacitor.ionicframework.com/docs/cordova/using-cordova-plugins/

```
npm install @ionic-native/media
npm install cordova-plugin-media
npx cap sync
```

I am then including the plugin within my page script:

```
import { Plugins } from '@capacitor/core';
import { Media, MediaObject } from '@ionic-native/media';
...
const Library: React.FC = () => {
  const playAudio = async () => {
    
  console.log("play request triggered");
    // const media_item = Media.create(testaudio);
    const testaudio = "https://link-to-audio-file.mp3";
    const file: MediaObject = Media.create(testaudio); 

    // to listen to plugin events:
    file.onStatusUpdate.subscribe(status => console.log(status)); // fires when file status changes

    file.onSuccess.subscribe(() => console.log('Action is successful'));

    file.onError.subscribe(error => console.log('Error!', error));

    // play the file
    file.play();
    console.log(file);


  }
// code not related to the issue
  const [items, setItems] = React.useState([]);
  React.useEffect(() => {
    sendGetRequest().then(
      data => setItems(data.products)
      );

  }, []);

  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonTitle>My Music</IonTitle>
        </IonToolbar>
      </IonHeader>
      <IonContent>
      <IonRow>
        {
            items.map(item => {

              return (
              <IonCol size="12">
                <IonCard>
                  <IonImg src={item['info']['thumbnail']} />
                  <IonCardHeader>
                    <IonCardSubtitle>Audio Length</IonCardSubtitle>
                    <IonCardTitle>{item['info']['title']}</IonCardTitle>
                  </IonCardHeader>
                  <IonCardContent>
                    {item['info']['content']}
                    
                  </IonCardContent>

                  <IonItem routerLink="/library/details" >
                      <IonIcon icon={play} slot="start" onClick={playAudio}/>
                      <IonLabel>
                        <h2>Session starten</h2>
                      </IonLabel>
                    </IonItem>
                    <IonItem routerLink="/library/details">
                      <IonIcon icon={heart} slot="start" />
                      <IonLabel>
                        Für später merken
                      </IonLabel>
                    </IonItem>
                 </IonCard>
                </IonCol>
              );
            })
          }
        
        </IonRow>
     
      </IonContent>
    </IonPage>
  );
};

```

 Most demo code using the plugin in Ionic seems to be using Angular. I'm aware that React is still a new platform for Ionic so I'm wondering if I'm doing anything wrong or maybe it is still to early on to use it for my use case?

Any hints are much appreciated.

**Comment by Deboracgs** [comment]
> I am currently trying to get the plug-in to work with capacitor and Ionic react.
> 
> First, I've installed the plug-in as suggested running the following commands described on this help page: https://capacitor.ionicframework.com/docs/cordova/using-cordova-plugins/
> 
> ```
> npm install @ionic-native/media
> npm install cordova-plugin-media
> npx cap sync
> ```
> 
> I am then including the plugin within my page script:
> 
> ```
> import { Plugins } from '@capacitor/core';
> import { Media, MediaObject } from '@ionic-native/media';
> ...
> const Library: React.FC = () => {
>   const playAudio = async () => {
>     
>   console.log("play request triggered");
>     // const media_item = Media.create(testaudio);
>     const testaudio = "https://link-to-audio-file.mp3";
>     const file: MediaObject = Media.create(testaudio); 
> 
>     // to listen to plugin events:
>     file.onStatusUpdate.subscribe(status => console.log(status)); // fires when file status changes
> 
>     file.onSuccess.subscribe(() => console.log('Action is successful'));
> 
>     file.onError.subscribe(error => console.log('Error!', error));
> 
>     // play the file
>     file.play();
>     console.log(file);
> 
> 
>   }
> // code not related to the issue
>   const [items, setItems] = React.useState([]);
>   React.useEffect(() => {
>     sendGetRequest().then(
>       data => setItems(data.products)
>       );
> 
>   }, []);
> 
>   return (
>     <IonPage>
>       <IonHeader>
>         <IonToolbar>
>           <IonTitle>My Music</IonTitle>
>         </IonToolbar>
>       </IonHeader>
>       <IonContent>
>       <IonRow>
>         {
>             items.map(item => {
> 
>               return (
>               <IonCol size="12">
>                 <IonCard>
>                   <IonImg src={item['info']['thumbnail']} />
>                   <IonCardHeader>
>                     <IonCardSubtitle>Audio Length</IonCardSubtitle>
>                     <IonCardTitle>{item['info']['title']}</IonCardTitle>
>                   </IonCardHeader>
>                   <IonCardContent>
>                     {item['info']['content']}
>                     
>                   </IonCardContent>
> 
>                   <IonItem routerLink="/library/details" >
>                       <IonIcon icon={play} slot="start" onClick={playAudio}/>
>                       <IonLabel>
>                         <h2>Session starten</h2>
>                       </IonLabel>
>                     </IonItem>
>                     <IonItem routerLink="/library/details">
>                       <IonIcon icon={heart} slot="start" />
>                       <IonLabel>
>                         Für später merken
>                       </IonLabel>
>                     </IonItem>
>                  </IonCard>
>                 </IonCol>
>               );
>             })
>           }
>         
>         </IonRow>
>      
>       </IonContent>
>     </IonPage>
>   );
> };
> ```
> 
> Most demo code using the plugin in Ionic seems to be using Angular. I'm aware that React is still a new platform for Ionic so I'm wondering if I'm doing anything wrong or maybe it is still to early on to use it for my use case?
> 
> Any hints are much appreciated.

I got worked this in ios, but on android doesn't work. I using react too.

`V/Capacitor/Plugin: To native (Cordova plugin): callbackId: INVALID, service: Media, action: startPlayingAudio, actionArgs: ["5066881b-f3cd-1584-329f-6074410a385c","teste.mp3",null]
W/AudioManager: Use of stream types is deprecated for operations other than volume control
    See the documentation of requestAudioFocus() for what to use instead with android.media.AudioAttributes to qualify your playback use case`

**Comment by d9media** [comment]
Thank you for your input Deboracgs!

I ended up moving away from Ionic for the current project after reviewing the state of capacitor vs. cordova plugin, the lack of documentation and overall community activity I no longer have confidence in the technology. I managed to get most of my requirements done with Flutter which has much better documentation too.

I am still curios though, how did you end up implementing the code on the React side? Maybe this will help future users understand the implementation better. Cheers!

**Comment by Deboracgs** [comment]
> Thank you for your input Deboracgs!
> 
> I ended up moving away from Ionic for the current project after reviewing the state of capacitor vs. cordova plugin, the lack of documentation and overall community activity I no longer have confidence in the technology. I managed to get most of my requirements done with Flutter which has much better documentation too.
> 
> I am still curios though, how did you end up implementing the code on the React side? Maybe this will help future users understand the implementation better. Cheers!

Hello!!  below is my code, I'm using typescript

```
import React from 'react';
import { connect } from "react-redux";
import { withRouter, RouteComponentProps } from 'react-router-dom'
import { BackgroundImage, BackgroundCirclesImage } from './../../components';
import { } from './../../actions';

import {  Media, MediaObject, MediaOriginal  } from '@ionic-native/media';
import { IonRange, IonIcon, IonButton, IonImg, isPlatform, getPlatforms } from '@ionic/react';




interface Props extends RouteComponentProps {

}



class PlayerPage extends React.Component<Props>{

    sound: MediaObject;

    constructor(props: Props, sound: MediaObject) {
        super(props);
        
       this.sound = sound;

    }

    componentDidMount(){
        this.sound = Media.create("https://www.alertweb.tech/assets/teste.mp3");
    }

    play(){

        this.sound.getCurrentPosition().then((value) => console.log("current position play ", value))
        this.sound.play({playAudioWhenScreenIsLocked: true, numberOfLoops: 1})
    }

    pause(){
        this.sound.getCurrentPosition().then((value) => console.log("current position pause ", value))
        this.sound.pause()
    }

    next(){
        
    }

    back(){

    }

    render() {
        return (
            <BackgroundImage gradient="180deg, #230640 0%, #230640 100%" image={BackgroundCirclesImage}>

                <div className="container player-page">
                    <div className="row ">
                        <div className="col s12">
                           <IonImg src={"https://images.prismic.io/ionicframeworkcom/fc6f762d-ebe2-4ad7-8746-233c36568907_Group+35.png?auto=compress,format&rect=0,0,280,200&w=280&h=200"} style={{ "width": "100%", "height": 200 }} />
                        </div>
                    </div>

                    <div className="row ">
                        <div className="col s12">
                            <IonButton color="secondary">back</IonButton>
                            <IonButton color="secondary" onClick={this.play.bind(this)}>play</IonButton>
                            <IonButton color="secondary" onClick={this.pause.bind(this)}>pause</IonButton>
                            <IonButton color="secondary">next</IonButton>
                        </div>
                    </div>

                    <div className="row ">
                        <div className="col s12">
                            <IonRange min={20} max={80} step={2} />
                        </div>
                    </div>
                </div>
                
            </BackgroundImage>
        );
    }

}


const mapStateToProps = ({ }) => {
    return {};
}
export default withRouter(connect(mapStateToProps, {

})(PlayerPage));
```

**Comment by d9media** [comment]
Makes sense. Thanks.

Hopefully you will be able to make it work on Android too.

**Comment by johnsonzsong** [comment]
@Deboracgs Thanks for the example! I'm also running into an issue with using this on Android and wanted to know how you ended up getting audio to work in Android? Thank you!

**Comment by timbru31** [comment]
I'm closing this issue since we can't give support for Capacitor and general support should be conducted via [StackOverflow](https://stackoverflow.com/questions/tagged/cordova) or in our [Slack community](http://slack.cordova.io). If you can reproduce the issue on Android in a new, plain Cordova, not Capacitor, app, please open a new issue for this.

More information on how to create a minimal reproduction repository: https://github.com/apache/cordova-contribute/blob/master/create-reproduction.md

**Comment by Deboracgs** [comment]
> @Deboracgs Thanks for the example! I'm also running into an issue with using this on Android and wanted to know how you ended up getting audio to work in Android? Thank you!

Hello, I didn't finish that project, but i thinking need to work around native modules.

**Comment by sefrem** [comment]
@johnsonzsong 
Hey. Not sure if you still need this but here is how i handled this case in React.
Here is my helper function 

```
export const useSound = (name) => {
  const path = File.applicationDirectory + `public/static/sounds/${name}.mp3`;
  let sound;
  if (Media) {
    sound = Media.create(path);
  }

  sound.onStatusUpdate.subscribe(status => {
    if (status === 4) sound.release();
  })

  return { sound }
}
```

Then in the component i just pass the name of the sound i want to play and play it like so

```
const { sound } = useSound('Example');
sound.play();
```
One thing that gave me the most pain is setting the proper file path. Here is how File.applicationDirectory works - https://android.googlesource.com/platform/frameworks/base.git/+/android-4.2.2_r1/core/java/android/webkit/URLUtil.java

> to refer to bar.png under your package’s asset/foo/ directory, use
>  “file:///android_asset/foo/bar.png”.

So basically File.applicationDirectory gives you the path to the assets directory in your project and then you go from there.

**Comment by johnsonzsong** [comment]
@sefrem Thanks for the example of how you handled it!
